### PR TITLE
Support compare in MongoDB

### PR DESCRIPTION
**Pull request by lennartvdd**
Do not merge yet - Depends on PR yiisoft/yii2#8505

Related to issue yiisoft/yii2#2022
Related to issue yiisoft/yii2#2315 inspired by PR yiisoft/yii2#4552

### CODE REVIEW

File: Collection.php [hunk 1]
```diff
@@ -844,21 +844,45 @@ public function buildCondition($condition)
             return [];
         }
         if (isset($condition[0])) { // operator format: operator, operand 1, operand 2, ...
-            $operator = strtoupper($condition[0]);
-            if (isset($builders[$operator])) {
```
Comment (klimov-paul) [inline Collection.php]: What is this change for?

File: Collection.php [hunk 2]
```diff
@@ -844,21 +844,45 @@ public function buildCondition($condition)
...
+    
+    /**
+     * Creates a condition on a column-value pair using an abritrary operator
+     * 
+     * @param string $operator The Mongo operator. Note that Mongo query operators must start with an $ (e.g. `$gt`, `$lt`, etc.)
+     * @param array $operands The column-value pair
+     * @return array The generated Mongo condition
+     * @throws InvalidParamException if $operator does not start with $ or when wrong number of operands have been given.
+     */
+    public function buildSimpleCondition($operator, $operands)
```
Comment (klimov-paul) [inline Collection.php]: Why this method introduced?

Comment (lennartvdd) [inline Collection.php]: This change is based on PR yiisoft/yii2#4552 's attempt to fix issue yiisoft/yii2#2315, which did not seem to be ported to the mongodb extension. 

Please also see my last comment in PR yiisoft/yii2#8505 

If there's a better way I'm happy to hear it.

Comment (klimov-paul) [inline Collection.php]: This does not justify extra method creation.

Comment (lennartvdd) [inline Collection.php]: Why not? The buildCondition method is a gateway method that calls specific conditionBuilders. Introducing the simple condition logic straight away into that method is not following single responsibility imho. But if it makes you happy I'll put the logic immediately in that method. Idk.  Let me know what you want and I'll change it.

Comment (klimov-paul) [inline Collection.php]: That is excatly the point. Original set of method was enough to compose all kind of MongoDB conditions, including comparisions.
That is why I am asking, why the extra method is need for this feature?

Comment (lennartvdd) [inline Collection.php]: I don't really get what's against adding a method. If you think this method is too simple to justify being in it's own descriptively named method, then how does it compare to the `buildLikeCondition()` for example? Why does that deserve it's own method and this doesn't? And the `buildAndCondition()` and `buildOrCondition()`? They contain _exactly_ the same code..

Comment (lennartvdd) [inline Collection.php]: Oh I guess I misunderstood you. Well. the current functionality does not allow for yii\db\QueryBuilder's syntax of  calling the querybuilder like this:  `$collection->buildCondition(['$gt', 'column', $value])`. This change allows for this.

Comment (klimov-paul) [inline Collection.php]: Such abillity is not provided because it is not suitable for MongoDB.

Comment (lennartvdd) [inline Collection.php]: Why not?

Comment (lennartvdd) [inline Collection.php]: The extra method has a valid use case imo. The extension becomes more intuitive to use the more it's behavior mimics `yii\db` behavior. The extra method enables you to use syntax that's already available `yii\db`, but currently not in `yii\mongodb`. 

But perhaps there's another reason why adding this method is a problem?

**Comment by klimov-paul** [comment]
#36

**Comment by lennartvdd** [comment]
Much cleaner this way. Thanks!

**Comment by klimov-paul** [comment]
Merged. Thank you.

**Comment by klimov-paul** [comment]
Adjusted at 1f5595331b7ead1106b7417dd0e12761a5ec8dc2

**Pull request merged**